### ISSUE TITLE
DNS plugin segfaults - OpenWRT / kernel 4.14

### ISSUE THREAD

**Issue by Strykar** [collectd/collectd#3796]
*   Version of collectd: `5.12`
*   Operating system / distribution: `OpenWrt 19.07.4 r11208-ce6496d796`
*   Kernel version (if applicable): `4.14.195 #0 SMP x86/64 Linux`

## Actual behavior
Given enough time running, the DNS plugin will segfault and crash, shutting down collectd.
```
Sat Dec 26 14:14:48 2020 daemon.err collectd[17396]: dns plugin: pcap_loop exited with status -1.
Sat Dec 26 14:15:13 2020 kern.info kernel: [103791.357292] reader#0[17408]: segfault at 7f4142f1a758 ip 00007f414313519a sp 00007f4142f1a760 error 6 in dns.so[7f4143135000+2000]
Sat Dec 26 14:15:13 2020 daemon.info procd: Instance collectd::instance1 s in a crash loop 6 crashes, 271 seconds since last crash
```
I have opened https://github.com/openwrt/packages/issues/14339, but the maintainer suggested this should be reported upstream.

This behavior has been going on for a long time even in previous versions on OpenWRT.
The OpenWRT package maintainer tried to help debug it in the past but that did not appear to go anywhere and the issue was closed. See https://github.com/collectd/collectd/issues/1227 the issue seems to be related to the DNS plugin using libpcap.

Issue is not platform specific (I have know of this issue for years on x64) and the maintainer was troubleshooting this on ar71xx / mips32r2 architectures.

The current workaround suggested for OpenWRT is to [disable the plugin](https://forum.openwrt.org/t/collectd-dns-plugin-crashes/20386/2) as its crashing causes collectd to exit.

## Steps to reproduce

*   Install and enable collectd on OpenWRT.
*   Wait 10 - 90 mins
*   View syslog to confirm collectd has crashed

While I could provide debug logs, I am not half as competent as @hnyman and his attempts in https://github.com/collectd/collectd/issues/1227
Perhaps OpenWRT could be added to Collectd's test suite as this issue is easily reproducible?

**Comment by wulfy23** [comment]
Not sure if exactly related but may as well list here...

On master @ rt171xx ... lack of ipv6 causes either startup or periodic crash issues... (runs mostly fine for me with ipv6...) 

collectd - 5.12.0-9

**Comment by jivanpal** [comment]
I'm experiencing this issue with the same version of collectd (5.12.0) on OpenWrt 23.05.2, Linux kernel 5.15.137. Willing to provide relevant logs if some guidance is given.

**Comment by hnyman** [comment]
I don't think that much has changed since my extensive debugging in 2015, documented on #1227 
(The only change is that there is even more variations to embedded device CPUs)

The normal system logs are useless in debugging this issue.  If you really want to debug, you can continue deeper from my debug expriments.

**Comment by eero-t** [comment]
@hnyman Isn't your workaround any more applied to OpenWRT collectd version, or is it not effective any more: https://github.com/collectd/collectd/issues/1227#issuecomment-139247935 ?

> The normal system logs are useless in debugging this issue. If you really want to debug, you can continue deeper from my debug expriments.

There are few additional tools that you could try:
* Building collectd with AddressSanitizer enabled: https://github.com/google/sanitizers/wiki/AddressSanitizer
* Using Valgrind memory checker: https://valgrind.org/docs/manual/mc-manual.html
* Using Valgrind threading / data race checkers:
  - https://valgrind.org/docs/manual/hg-manual.html
  - https://valgrind.org/docs/manual/drd-manual.html

They are all fairly easy to use, first one requires just few compile options, and Valgrind is CLI tool you use to prefix collectd command.

(Checking the reported potential threading race issues is where the actual work is.)

**Comment by itsthejb** [comment]
Now seeing this regularly on x86 `OpenWrt 24.10.4 r28959-29397011cc`. Which is strange because I didn't encounter it for years and also not on x86 for about a year
